resources/files/windows/PuppetBolt/PuppetBoltFunctions.ps1

```diff
@@ -0,0 +1,1639 @@
+
+        [parameter(ValueFromPipeline=$True,ValueFromPipelineByPropertyName)]
+        [BoltOutputLevel]  $OutputLevel
+    )
+    begin {
+        $_flags = Convert-BoltParams @PSBoundParameters
+    }
+    process {
+        if ($OutputLevel) {
+            $_result = Invoke-Expression "bolt task show $TaskName --format=json $_flags"
+        } else {
+            $_result = Invoke-Expression "bolt task show $TaskName --format=json $_flags" 2>$null
+        }
+        return ($_result | ConvertFrom-Json)
+    }
+}
+
+function Read-BoltModules {
+    <#
+    .SYNOPSIS
+    Read all modules available to Puppet Bolt
+    .DESCRIPTION
+    This function will execute 'bolt puppetfile show-modules' to list available modules.
+    Puppet Bolt (https://puppet.com/products/bolt) is an agentless automation
+    solution for running ad-hoc tasks and operations on remote targets
+    .PARAMETER ModulePath
+    List of directories containing modules, separated by ';'
+    .PARAMETER BoltDir
+    Specify what Boltdir to load config from (default: autodiscovered from current working dir)
+    .PARAMETER ConfigFile
+    Specify where to load config from (default: ~/.puppetlabs/bolt/bolt.yaml)
+    .PARAMETER Color
+    Whether to show output in color
+    .PARAMETER OutputLevel
+    Level of output logging to display (defaults to no output). Must be one of:
+    * 'Verbose' only verbose logging
+    * 'Debug' both verbose and debug logging
+    * 'Trace' all of verbose, debug, and stacktrace logging
+    #>
+    [CmdletBinding()]
+    param(
+        [parameter(ValueFromPipeline=$True,ValueFromPipelineByPropertyName)]
+        [string]  $ModulePath,
+
+        [parameter(ValueFromPipeline=$True,ValueFromPipelineByPropertyName)]
+        [string]  $BoltDir,
+
+        [parameter(ValueFromPipeline=$True,ValueFromPipelineByPropertyName)]
+        [string]  $ConfigFile,
+
+        [parameter(ValueFromPipeline=$True,ValueFromPipelineByPropertyName)]
+        [bool]    $Color=$true,
+
+        [parameter(ValueFromPipeline=$True,ValueFromPipelineByPropertyName)]
+        [BoltOutputLevel]  $OutputLevel
+    )
+    begin {
+        $_flags = Convert-BoltParams @PSBoundParameters
+    }
+    process {
+        if ($OutputLevel) {
+            $_result = Invoke-Expression "bolt puppetfile show-modules --format=json $_flags"
+        } else {
+            $_result = Invoke-Expression "bolt puppetfile show-modules --format=json $_flags" 2>$null
+        }
+        return ($_result | ConvertFrom-Json)
+    }
+}
+
+function Install-BoltModules {
+    <#
+    .SYNOPSIS
+    Install bolt modules from puppetfile
+    .DESCRIPTION
+    This function will execute 'bolt puppetfile install' to install any modules required
+    in the puppetfile. Puppet Bolt (https://puppet.com/products/bolt) is an agentless automation
+    solution for running ad-hoc tasks and operations on remote targets
+    .PARAMETER ModulePath
+    List of directories containing modules, separated by ';'
```

Comment:
This is not idomatic PowerShell.  It should read this one item at a time from the pipeline instead.

Reply:
I agree with the principle, and I've changed some of the other commandlets, but I think in the case of ModulePath we _don't_ want the commandlet to do that.

As a user I wouldn't expect setting ModulePath to make the commadlet iterate over each path. ModulePath should be akin to a load path, so I expect the command to execute with the whole ModulePath, and include all the directories available in ModulePath. Otherwise each iteration would only include one of the possibly multiple paths for loading which would lead to unintended consequences (for example, if you have a module which loads a module in another modulepath it would break if we iterated over each path one at a time.)